app/src/main/java/com/android/sample/ui/filter/FilterPage.kt

```diff
@@ -0,0 +1,317 @@
+  Column(
+      modifier =
+          Modifier.padding(paddingValues).fillMaxSize().verticalScroll(rememberScrollState())) {
+        ValueRangeSlider(
+            modifier = Modifier.testTag("timeRangeSlider"),
+            name = TIME_RANGE_NAME,
+            min = TIME_RANGE_MIN,
+            max = TIME_RANGE_MAX,
+            unit = TIME_RANGE_UNIT,
+            range = recipesViewModel.filter.value.timeRange,
+            updateRange = { newMin, newMax -> recipesViewModel.updateTimeRange(newMin, newMax) })
+        ValueRangeSlider(
+            modifier = Modifier.testTag("priceRangeSlider"),
+            name = PRICE_RANGE_NAME,
+            min = PRICE_RANGE_MIN,
+            max = PRICE_RANGE_MAX,
+            unit = PRICE_RANGE_UNIT,
+            range = filter.priceRange,
+            updateRange = { newMin, newMax -> recipesViewModel.updatePriceRange(newMin, newMax) })
+        CheckboxDifficulty(recipesViewModel = recipesViewModel)
+        CheckboxCategories(recipesViewModel = recipesViewModel)
+      }
+}
+
+/**
+ * Composable function to display a range slider.
+ *
+ * @param name The name of the slider.
+ * @param min The minimum value of the slider.
+ * @param max The maximum value of the slider.
+ * @param unit The unit of the slider.
+ * @param range The range of the slider.
+ * @param updateRange The function to update the range.
+ */
+@Composable
+fun ValueRangeSlider(
+    modifier: Modifier,
+    name: String,
+    min: Float,
+    max: Float,
+    unit: String,
+    range: FloatRange,
+    updateRange: (Float, Float) -> Unit,
+) {
+  // Update the range to the [min, max] if it is unbounded
+  if (range.isLimited()) {
+    range.update(min, max)
+  }
+  // Mutable state to store the slider position
+
+  var sliderPosition by remember { mutableStateOf(range.min..range.max) }
+  Column(modifier = Modifier.padding(16.dp), horizontalAlignment = Alignment.CenterHorizontally) {
+    Text(
+        text = name,
+        style = MaterialTheme.typography.titleMedium,
+        modifier = Modifier.align(Alignment.Start))
+
+    Spacer(modifier = Modifier.height(8.dp))
+
+    // Display selected min and max values
+    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
+      Text(
+          "${sliderPosition.start.toInt()} " + unit,
+          style = MaterialTheme.typography.bodyMedium,
+          color = MaterialTheme.colorScheme.onPrimary)
+      Text(
+          "${sliderPosition.endInclusive.toInt()} " + unit,
+          style = MaterialTheme.typography.bodyMedium,
+          color = MaterialTheme.colorScheme.onPrimary)
+    }
+
+    Spacer(modifier = Modifier.height(8.dp))
+
+    // Range slider with minimum 0 and maximum 120 minutes
+    RangeSlider(
+        modifier = modifier.fillMaxWidth(),
+        value = sliderPosition,
+        onValueChange = {
+          sliderPosition = it
+          updateRange(it.start, it.endInclusive)
```

Comment:
From my understanding, each time the user adjusts the slider, this triggers `recipesViewModel.updatePriceRange(newMin, newMax)`, which currently just updates `_filter = MutableStateFlow(Filter())`. However, if updating this value eventually triggers a new fetch, we could end up with a significant number of unnecessary API calls. Since from my understanting this part is not implemented we need to be carefull.

For optimization, it might be more efficient to introduce a 'Save' button that applies all selected filters at once, reducing the API requests and improving performance.

Reply:
Thank you for the detailed feedback! You’re correct that, as it stands, each slider adjustment updates ? 
`recipesViewModel.updatePriceRange(newMin, newMax)`. Once the backend is available, I’ll optimize this within `RecipeViewModel` to minimize API calls by using a range of values that only triggers updates when necessary.

Regarding the 'Save' button, I actually prefer not to include it, as the main goal of my UI design is to keep interactions as simple as possible. After reviewing similar applications, it seems that saving changes automatically (rather than using a button) creates a smoother user experience, especially since users can directly delete filters from the Swipe Page. For now, I’ll leave the current setup as is, and when the backend is complete, I’ll handle filtering in a way that’s dependent on the database.